packages/flutter/lib/src/widgets/overscroll_indicator.dart

```diff
@@ -834,138 +826,223 @@ class _StretchingOverscrollIndicatorState extends State<StretchingOverscrollIndi
+  /// Manages and notifies changes to the current [overscroll] value.
+  final ValueNotifier<double> _overscrollNotifier = ValueNotifier<double>(0.0);
+  double get overscroll => _overscrollNotifier.value;
+  set overscroll(double newValue) {
+    _overscrollNotifier.value = clampDouble(newValue, minOverscroll, maxOverscroll);
+  }
+
+  /// Stores the `overscroll` value from an ongoing animation at the precise
+  /// moment it is interrupted by a new `pull()` gesture.
+  ///
+  /// When `pull()` is called while an animation (triggered by `absorbImpact`
+  /// or `scrollEnd`) is active, this field captures the animation's current
+  /// `_controller.value` (which represents the overscroll amount) immediately
+  /// before the animation controller is disposed.
+  ///
+  /// This captured value is then added to the overscroll amount calculated
+  /// by the `pull()` method. The primary purpose is to create a smoother
+  /// visual transition from an animated overscroll state to a direct,
+  /// user-driven pull, minimizing any abrupt visual "jumps" in the
+  /// stretch effect.
+  ///
+  /// It is reset to `0.0` when an animation completes naturally via `animate()`
+  /// or when a new pull starts without a preceding active animation.
+  double _interruptedOverscroll = 0.0;
 
   // Constants from Android.
   static const double _exponentialScalar = math.e / 0.33;
   static const double _stretchIntensity = 0.016;
-  static const double _flingFriction = 1.01;
-  static const Duration _stretchDuration = Duration(milliseconds: 400);
 
-  double get value => _stretchSize.value;
+  static const double minOverscroll = -1.0;
+  static const double maxOverscroll = 1.0;
 
-  // Constants for absorbImpact.
-  static const double _kMinVelocity = 1;
-  static const double _kMaxVelocity = 10000;
-  static const Duration _kMinStretchDuration = Duration(milliseconds: 50);
+  /// A fraction used to adjust the input velocity, measured in pixels,
+  /// to a value between -1 and 1 when gesture fling.
+  static const double _flingVelocityFriction = 1 / 6000;
+
+  /// The maximum velocity allowed for a fling after scaling
+  /// to prevent applying an excessive stretch effect.
+  static const double _maxFlingVelocity = 0.5;
+
+  @override
+  void addListener(VoidCallback listener) {
+    _overscrollNotifier.addListener(listener);
+  }
+
+  @override
+  void removeListener(VoidCallback listener) {
+    _overscrollNotifier.removeListener(listener);
+  }
+
+  /// Creates a stretching-only [Simulation] ported from Android 12.
+  _StretchSimulation _createStretchSimulation(double velocity) {
+    return _StretchSimulation(overscroll, velocity, Tolerance.defaultTolerance);
+  }
+
+  /// Uses the saturation function approach to scale velocity.
+  /// Small velocities remain roughly proportional, while large
+  /// velocities gradually saturate. Returns the scaled velocity
+  /// with the original direction preserved.
+  double _calculateVelocityScale(double velocity) {
```

Comment:
Can you explain where this part comes from? I can't find similar code from either `EdgeEffect` or the code before the PR.